Improve performance of retrieving blob data

betesh opened the pull request:
I discovered that Porcelain.grep performs horribly.  I did some profiling and identified a few issues, mostly involving unnecessarily retrieving data that we already had in memory.

I tested performance against a repo of mine and a085d06bcc5ddd6880de479250f55586cfae0935 improved runtime of a simple `Porcelain.grep` call from 6.1 seconds to 0.57 seconds.

Commit d5019e6a67453a0191408067bcb9c53436c9ee25 further improved runtime to 0.28 seconds.  (The performance bottleneck that d5019e6a67453a0191408067bcb9c53436c9ee25 bypasses is `Java::OrgEclipseJgitTreewalk::TreeWalk.forPath`.)

The last remaining bottleneck was `get_bytes.to_a`, accounting for 80% of runtime.  b3d01eca9f3b802178ba77646ed5c46e05743623 fixes this.  I ran this against a much larger repo, and b3d01eca9f3b802178ba77646ed5c46e05743623 alone cut runtime from 103 seconds to 7 seconds!

betesh commented:
@bartkamphorst Please take a look--this PR improves performance of `Porcelain.grep` by 99.7%

bartkamphorst commented:
Impressive speed-up @betesh ! Unfortunately I won't have time to review this PR this week. I'll try to get to it after the weekend (or @dometto ?).

betesh commented:
@bartkamphorst Reminder

betesh commented:
@bartkamphorst ?

bartkamphorst commented on lib/rjgit.rb:
```diff
@@ -209,15 +216,32 @@ def self.grep(repository, query, options={})
-        next if blob.binary?
+      # We optimize below by first grepping the entire file, and then, if a match is found, then identifying the individual line.
+      # To avoid missing full-line matches during the optimization, we first convert multiline anchors to single-line anchors.
+      query = Regexp.new(query.source.gsub(/\A\\A/, '^').gsub(/\\z\z/, '$'), query.options)
```
I'm not quite sure why this step would be necessary; can you explain why we would miss full-line matches without this step?

betesh replied:
From the user's perspective, we are searching each line individually for a match, the same way `git grep` works.  So a user would reasonably expect that `\A` and `\z` should refer to the beginning and end of an individual line in the file, not to the beginning and end of the entire file.

However, because we optimize below by first searching the entire file (`next unless query.match(file_contents)`), `\A` and `\z` mean the beginning and end of the entire file during that match.

For example, if a user greps for `\A123`, they mean "Any line that begins with 123".  If the first line of the file begins with `456` and the second line begins with `123`, `next unless query.match(file_contents)` will omit that file from results.  By converting to `^123`, we ensure that the 2nd line of the file matches both when we search the entire file and when we later search line-by-line.

dometto replied:
Thanks for the explanation!

bartkamphorst commented on lib/rjgit.rb:
```diff
@@ -209,15 +216,32 @@ def self.grep(repository, query, options={})
+        binary = RawText.is_binary(bytes)
+        next if binary
+
+        file_contents = bytes.to_s
```
I'm surprised that this doesn't mess up any of the tests! Happy with the resulting speed-up though. :)

dometto commented on lib/rjgit.rb:
```diff
@@ -55,11 +55,18 @@ def self.object_for_tag(repository, tag)
 
     # http://dev.eclipse.org/mhonarc/lists/jgit-dev/msg00558.html
-    def self.cat_file(repository, blob)
+    def self.cat_file(repository, blob, path = nil)
```
I like that idea!

betesh replied:
Done: c7084b5dff1df50f71cf44a4c7d952485a900c2b

dometto commented on lib/rjgit.rb:
```diff
@@ -55,11 +55,18 @@ def self.object_for_tag(repository, tag)
       # Try to resolve symlinks; return nil otherwise
-      mode = RJGit.get_file_mode(jrepo, jblob)
+      mode = if path
+        last_commit_hash = jrepo.resolve(Constants::HEAD)
```
This mirrors our code [here](https://github.com/repotag/rjgit/blob/master/lib/blob.rb#L75), where we are, however, allowing the user to choose the `revstring` as a parameter. The more fundamental issue is the following.

Git file modes are not a property of a blob as such. The same blob can technically have a different file mode in different trees (and therefore in different branches). For `cat_file` purposes, we just want to know if a blob is a symlink, so that we can cat the contents of the linked file, rather than the contents of the symlink-blob. The contents of a symlink-blob will consist of only the path to the linked file. So the question is: relative to what branch do we want to determine whether a blob (in a given tree `path`) is a symlink or not?

I think it makes sense to make `HEAD` (= the 'active' branch) the default here, as `cat_file` is a `Porcelain` method. We are, however, also using this `Porcelain` method to cat the contents of `Blob`s "plumbing style". In that context, 'active' branch doesn't really have a meaning. Here are two options:

* default to`HEAD`, but let the user specify an optional revstring in `Blob#data` and `Porcelain.cat_file`.
* let `Porcelain.cat_file` accept an optional `mode` argument (default: `nil`), and an optional `lookup_mode` argument (default: `true`). In `Blob#data`, call `cat_file` with the blob's mode if the user has set it (or `nil` otherwise) and `lookup_mode = false`.

betesh replied:
@dometto In the `grep` code path, which is the only use case that requires looking up the mode based on path here, the blob already knows its mode.  I think we can drop the "lookup mode from path" piece in favor of passing the mode into this method.  I've implemented this in ceb3c06a37089a9c3cd00c0fb343eeec4a753584, removing most of the changes I introduced to this method.

dometto replied:
Good solution, given that `grep` is indeed the only (plumbing) path for which this is an issue. When calling `cat_file` directly (porcelain api), the user will just have to supply the mode themselves. :+1:

dometto reviewed (commented):
Thanks for the contribution @betesh! Small comment in relation to one of @bartkamphorst's questions

betesh commented on lib/rjgit.rb:
```diff
@@ -55,11 +55,18 @@ def self.object_for_tag(repository, tag)
-      mode = RJGit.get_file_mode(jrepo, jblob)
+      mode = if path
+        last_commit_hash = jrepo.resolve(Constants::HEAD)
+        return nil if last_commit_hash.nil?
```
Fixed: 33740be84ae5782eb799415f3a3e88b2e968b19f

betesh commented:
I ran `Porcelain.grep` on a large repo (43619 files) to get some new metrics.

On master: 4138.94 seconds
On this branch (both before and after addressing PR feedback): 10.55 seconds

That's a 99.75% improvement

betesh commented:
@bartkamphorst  Sorry it took me so long to get to this.  I believe I've addressed all your feedback.

dometto commented:
> On master: 4138.94 seconds (i.e. 1 hour, 9 minutes)
> On this branch (both before and after addressing PR feedback): 10.55 seconds
> That's a 99.75% improvement

Amazing result! Thanks so much for working on this.

dometto commented on lib/rjgit.rb:
```diff
@@ -209,15 +210,38 @@ def self.grep(repository, query, options={})
+        bytes = repo.open(id).get_bytes
+
+        binary = RawText.is_binary(bytes)
+        next if binary
```
Perhaps just `next if RawText.is_binary(bytes)`, to avoid instantiating another object. We're not using `binary` anywhere else, right?

bartkamphorst replied:
Seems like it.

betesh replied:
Fixed: d7b6c75e1c6fa7ad7dc6a30c0e1740ab8b773b2a

dometto reviewed (commented):
One very minor comment from me. @bartkamphorst let us know what you think

bartkamphorst commented on lib/rjgit.rb:
```diff
@@ -209,15 +210,38 @@ def self.grep(repository, query, options={})
+          unless symlink_source[File::SEPARATOR]
+            dir = file[:path].split(File::SEPARATOR)
+            dir[-1] = symlink_source
+            symlink_source = File.join(dir)
```
What's happening here when you reassign `symlink_source`?

betesh replied:
Commit message says "Resolve source files of relative symlinks correctly".  From what I remember, it's because symlinks could be absolute or relative, but to open it we need an absolute path.  If there's no `/` in the path, that tells us it's relative, so we need to join it with the dirname to get the full path to it.

bartkamphorst replied:
Thanks for explaining. I guess I don't understand the reason for the intermediate `dir` array then. Why not instead do:
```ruby
symlink_source = ::File.join(file[:path], symlink_source) unless symlink_source.include?[::File::SEPARATOR]
```

bartkamphorst replied:
Thanks for explaining. I guess I don't understand the reason for the intermediate `dir` array then. Why not instead do:
```ruby
symlink_source = ::File.join(file[:path], symlink_source) unless symlink_source[::File::SEPARATOR]
```
(**EDIT** removed my `.include?` call for speed; your implementation of checking for '/' is faster.)

betesh replied:
`dir[-1]` (before I overwrite it) is the symlink file's name.  Given the following example:

```
$ ls -l /tmp/example
total 0
-rw-rw-r-- 1 my_user my_group 0 Nov 17 10:31 abc
lrwxrwxrwx 1 my_user my_group 3 Nov 17 10:31 def -> abc
```

`file[:path]` is '/tmp/example/def' and `symlink_source` is `abc`
`dir = file[:path].split(File::SEPARATOR)` is `['/tmp', 'example', 'def']`
`dir[-1] = symlink_source` changes `dir` to `['/tmp', 'example', 'abc']`
`symlink_source = File.join(dir)` results in '/tmp/example/abc'

Doing this in one line, as you propose, it would look like this:

```
symlink_source = file[:path].sub(/[^#{File::SEPARATOR}]+\z/, symlink_source) unless symlink_source[::File::SEPARATOR]
```

Personally, I find that less readable than the current implementation but I'll let you make the decision.

bartkamphorst replied:
Nope, I'm on board now! Thanks for helping me understand.

bartkamphorst reviewed (changes requested):
Great work! Impressive speedups, @betesh . Just one more question for my understanding, but code is looking good!

betesh commented:
@bartkamphorst Thanks for all the feedback.  Can you please merge & release?

bartkamphorst commented:
@betesh Done! Thanks again for all your work on this, and for your patience! 👍 If you come up with other improvements in the future, don't hesitate to let us know.
